mlflow/tensorflow.py

```diff
@@ -333,3 +343,125 @@ def predict(self, df):
+        mlflow.log_param('optimizer_name', self.model.optimizer.optimizer._name)
+        #TODO: Fix for keras log_model not saving TF optimizers
+        mlflow.keras.log_model(self.model, None, None)
+
+_metric_queue = []
+
+_thread_pool = concurrent.futures.ThreadPoolExecutor(max_workers=1)
+
+
+def flush_queue():
+    global _metric_queue
+    try:
+        client = mlflow.tracking.MlflowClient()
+        if mlflow.active_run() is not None:
+            client.log_batch(mlflow.active_run().info.run_id, metrics=_metric_queue, params=[], tags=[])
+    except Exception as e:
+        warnings.warn("Logging to MLflow failed: " + str(e))
+    finally:
+        _metric_queue = []
+
+
+atexit.register(flush_queue)
+
+
+def add_to_queue(key, value, step):
+    met = Metric(key=key, value=value, timestamp=int(time.time()*1000), step=step)
+    _metric_queue.append(met)
+    if len(_metric_queue) >= 500:
+        flush_queue()
+
+
+def _log_event(event):
+    if mlflow.active_run() is None:
+        mlflow.start_run()
+    if event.WhichOneof('what') == 'summary':
+        summary = event.summary
+        for v in summary.value:
+            if v.HasField('simple_value'):
+                if event.step % 100 == 0:
+                    _thread_pool.submit(add_to_queue, key=v.tag, value=v.simple_value, step=event.step)
+
+
+def _checkpoint_model(model_path):
+    mlflow.log_artifact(model_path)
+
+
+def autolog():
+    import tensorflow
+    from tensorflow.python.summary.writer.event_file_writer import EventFileWriter
+    from tensorflow.python.summary.writer.event_file_writer_v2 import EventFileWriterV2
+
+    def contains_tensorboard_callback(lst):
+        for x in lst:
+            if isinstance(x, tensorflow.keras.callbacks.TensorBoard):
+                return True
+
+    @gorilla.patch(tensorflow.estimator.Estimator)
+    def export_saved_model(self, *args, **kwargs):
+        from tensorflow.python.saved_model import tag_constants
+
+        original = gorilla.get_original_attribute(tensorflow.estimator.Estimator, 'export_saved_model')
+        serialized = original(self, *args, **kwargs)
+        log_model(tf_saved_model_dir=serialized.decode('utf-8'), tf_meta_graph_tags=[tag_constants.SERVING],
+                  tf_signature_def_key='predict',
+                  artifact_path='model')
+        return serialized
+
+
+    @gorilla.patch(tensorflow.keras.Model)
+    def fit(self, *args, **kwargs):
+        original = gorilla.get_original_attribute(tensorflow.keras.Model, 'fit')
+        if len(args) >= 6:
+            l = list(args)
+            if contains_tensorboard_callback(l):
+                l[5] = l[5] + [__MLflowTfKerasCallback()]
+            args = tuple(l)
+        elif 'callbacks' in kwargs:
+            if contains_tensorboard_callback(kwargs['callbacks']):
+                kwargs['callbacks'] = kwargs['callbacks'] + [__MLflowTfKerasCallback()]
```

Comment:
Sorry I forget, is this the right behavior because the TensorBoard callback doesn't use EventFileWriter (so we still need to add our custom logging callback)? Or will we double-log in this case?

Reply:
Yeah that's it, marking as resolved.

Reply:
I actually think we could just add a TensorBoard callback for the user if they don't have one, writing to some tempdir we compute (e.g. via `tempdir.mkdtemp()`) , so that users get metrics logged to MLflow even if they don't pass a TensorBoard callback (as @clemens-db pointed out, it is a big hurdle to have to add a TensorBoard callback). 

The pros are simpler migration from Keras -> tf.Keras & that we'd be much more likely to log stuff for tf.Keras users (vs them having to realize or read docs to realize they need to pass a TensorBoard callback).

The cons are maybe a very slight perf hit (just writing to files) & (more scary) potential failures if the TensorBoard callback throws, but I think it's pretty unlikely for it to throw. We can also get more people's opinions.

Reply:
I also wonder if we could log the TensorBoard files as an artifact of the MLflow run in this case (so users can view/download them from the UI, and maybe later we could add support for actually visualizing them with TensorBoard from the MLflow UI, or give them a command to do this)

Reply:
I agree with the first point. Logging params only when TensorBoard exists to handle metrics for us is probably not the most training-code-independent workflow we were aiming for, so we can add the TB callback.

I'm not sure logging the TB files as an artifact makes the most sense, because there's the natural question of where those files came from (or more importantly, what they are) for the users who didn't add in the TensorBoard callback. I think we can rely on TB as the source of TensorFlow data but not really as the end-goal.

Reply:
Discussed online - will log TB files as artifact. Leaving resolution to @smurching